Bind cauchy_, exponential_, normal_, uniform_ functions to THPVariable.

gchanan opened the pull request:
Also changes the error messages around Generator parser; previously, you'd get an error
like: torch._C.Generator is not a torch.Generator; now the check is proper but returns
that only None is supported.  We likely need to change THPGenerator to have an at::Generator rather than a THGenerator to support non-None.

gchanan committed:
Bind cauchy_, exponential_, normal_, uniform_ functions to THPVariable.

Also changes the error messages around Generator parser; previously, you'd get an error
like: torch._C.Generator is not a torch.Generator; now the check is proper but returns
that only None is supported.

apaszke commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -82,7 +82,7 @@ bool FunctionParameter::check(PyObject* obj) {
       return size > 0 && THPUtils_checkLong(obj);
     }
-    case ParameterType::GENERATOR: return false;
+    case ParameterType::GENERATOR: return THPGenerator_Check(obj);
```
Is there any point in doing this check if it can't be used anyway? `None`s should be handled in a different pathway anyway, right?

gchanan replied:
It's just a slightly nicer error message in the case where you actually pass in a generator.  I spent a few minutes trying to figure out why my torch.default_generator was a torch._C.Generator and not a torch.Generator before I figured out it just rejects any object and the error message is bogus.  Now it just tells you only None is supported in that case.

killeent reviewed (commented):
What is the difference between `# fallthrough` and `# reinforce` ?

gchanan commented:
those are just comments.

gchanan committed:
Support passing Generators to ATen Variable-bound methods.

This involves changing THPGenerator to have an at::Generator rather than a THGenerator.
TH getRNGState, setRNGState are still called directly because they are not bound from ATen yet;
they should probably be on the Generators and return (opaque) GenerateState objects.

gchanan commented:
Latest commit supports passing Generator objects through these methods;

This involves changing THPGenerator to have an at::Generator rather than a THGenerator.
TH getRNGState, setRNGState are still called directly because they are not bound from ATen yet;
they should probably be on the Generators and return (opaque) GenerateState objects rather than being bound through Tensor.

Also, THPGenerator_TH_CData is kind of an ugly looking macro but should go away once we want getRNGstate, setRNGState and merge Variable and Tensor.  LMK if you have a better name for it.

gchanan commented:
@pytorchbot retest this please.

gchanan committed:
Fix default values.

colesbury reviewed (approved):
lgtm!

colesbury commented:
`test_manual_seed` looks broken -- needs to be fixed before merge

gchanan committed:
Properly use THRandom_initialSeed.

gchanan committed:
update standard gamma to use new default generator.

pietern commented:
There was some CI maintenance happening this morning -- retriggering build.

@pytorchbot retest this please
